### PR TITLE
Kong TCPIngress translation

### LINKED ISSUE #85: [Kong provider] TCPIngress translation
**What would you like to be added**:

Kong Ingress controller leverages a CRD called `TCPIngress` to expose TCP routes. Such a resource should be converted into `Gateway` + `TCPRoute`s + `TLSRoute`s.

### PR DESCRIPTION
**Pull request by mlavacca**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #85 

Additionally, go has been bumped to 1.21

**Does this PR introduce a user-facing change?**:

```release-note
The Kong `TCPIngress` resources are properly translated into `Gateway`s, `TCPRoute`s, and `TLSRoute`s.
```

### CODE REVIEW

File: pkg/i2gw/providers/common/utils.go [hunk 1]
```diff
@@ -95,14 +95,18 @@ func NameFromHost(host string) string {
...
 	return step2
 }
 
 func RouteName(ingressName, host string) string {
-	return fmt.Sprintf("%s-%s", ingressName, NameFromHost(host))
+	routeName := ingressName
+	if host != "" {
+		routeName = fmt.Sprintf("%s-%s", ingressName, NameFromHost(host))
+	}
+	return routeName
```
Comment (LiorLieberman) [inline pkg/i2gw/providers/common/utils.go:109]: why?

File: pkg/i2gw/providers/kong/consts.go [hunk 2]
```diff
@@ -26,6 +31,23 @@ const (
 	pluginsKey = "plugins"
 )
 
+const (
+	v1beta1Version string = "v1beta1"
+
+	kongResourcesGroup gatewayv1beta1.Group = "configuration.konghq.com"
+
+	kongPluginKind gatewayv1beta1.Kind = "KongPlugin"
+	tcpIngressKind gatewayv1beta1.Kind = "TCPIngress"
```
Comment (LiorLieberman) [inline pkg/i2gw/providers/kong/consts.go:40]: why gatewayv1beta1 and not just strings or kubernetes v1 kinds/groups?

File: pkg/i2gw/providers/kong/kong.go [hunk 3]
```diff
@@ -51,25 +54,23 @@ func NewProvider(conf *i2gw.ProviderConf) i2gw.Provider {
 // ToGatewayAPI converts the received i2gw.InputResources to i2gw.GatewayResources
 // including the kong specific features.
 func (p *Provider) ToGatewayAPI(_ i2gw.InputResources) (i2gw.GatewayResources, field.ErrorList) {
-	return p.converter.convert(p.storage)
+	return p.convert(p.storage)
```
Comment (LiorLieberman) [inline pkg/i2gw/providers/kong/kong.go:57]: why moving a way from the common approach across the current implementations? 

We have a converter and the the converter is the one to have the method `convert`

File: pkg/i2gw/providers/kong/kong.go [hunk 4]
```diff
@@ -34,15 +34,18 @@ func init() {
...
+
+	*resourceReader
+	*converter
 }
 
 // NewProvider constructs and returns the kong implementation of i2gw.Provider.
 func NewProvider(conf *i2gw.ProviderConf) i2gw.Provider {
 	return &Provider{
-		storage:        newResourcesStorage(),
+		conf: conf,
```
Comment (LiorLieberman) [inline pkg/i2gw/providers/kong/kong.go:47]: I dont think we ever use this. 
the resourceReader is the one that holds the conf

File: pkg/i2gw/providers/kong/storage.go [hunk 5]
```diff
@@ -17,16 +17,19 @@ limitations under the License.
...
 )
 
 type storage struct {
-	Ingresses map[types.NamespacedName]*networkingv1.Ingress
+	Ingresses    map[types.NamespacedName]*networkingv1.Ingress
+	TCPIngresses []kongv1beta1.TCPIngress
 }
 
-func newResourcesStorage() *storage {
+func newStorage() *storage {
```
Comment (LiorLieberman) [inline pkg/i2gw/providers/kong/storage.go:30]: nit: any reason for having a different function name than other implementations?

File: pkg/i2gw/providers/kong/converter.go [hunk 6]
```diff
@@ -43,24 +45,37 @@ func newConverter() *converter {
 	}
 }
 
-func (c *converter) convert(storage *storage) (i2gw.GatewayResources, field.ErrorList) {
+func (c *converter) convert(storage storage) (i2gw.GatewayResources, field.ErrorList) {
```
Comment (LiorLieberman) [inline pkg/i2gw/providers/kong/converter.go:48]: nit: I think when you call convert you do pass a pointer to storage

File: go.mod
Comment (LiorLieberman) [inline go.mod:1]: why is this needed? can we revert this?

File: pkg/i2gw/providers/common/utils.go (hunk 1 above)
Comment (mlavacca) [reply pkg/i2gw/providers/common/utils.go:109]: because if there is no host set the resulting name is `<ingress-name>-`

Comment (LiorLieberman) [reply pkg/i2gw/providers/common/utils.go:109]: oh ok, isnt it simpler to just wrap it in trimsuffix?

Comment (mlavacca) [reply pkg/i2gw/providers/common/utils.go:109]: the conditional needs to be there in any case, hence I do not see any difference.

File: pkg/i2gw/providers/kong/converter.go (hunk 6 above)
Comment (mlavacca) [reply pkg/i2gw/providers/kong/converter.go:48]: nope, I don't

File: pkg/i2gw/providers/kong/storage.go (hunk 5 above)
Comment (mlavacca) [reply pkg/i2gw/providers/kong/storage.go]: Changed to be coherent with others.

File: pkg/i2gw/providers/kong/kong.go (hunk 4 above)
Comment (mlavacca) [reply pkg/i2gw/providers/kong/kong.go]: yep, removed

File: pkg/i2gw/providers/kong/kong.go (hunk 3 above)
Comment (mlavacca) [reply pkg/i2gw/providers/kong/kong.go]: no particular reason, there are equivalent. Reverted to be coherent with other implementations.

File: pkg/i2gw/providers/kong/consts.go (hunk 2 above)
Comment (mlavacca) [reply pkg/i2gw/providers/kong/consts.go]: it was more convenient as we use these constants in gateway `ObjectRef`s. I agree that `string` fits better though. Changed.

File: pkg/i2gw/providers/common/utils.go (hunk 1 above)
Comment (LiorLieberman) [reply pkg/i2gw/providers/common/utils.go:109]: so I am probably missing something, why does it need to be there?

File: go.mod
Comment (LiorLieberman) [reply go.mod:1]: go.mod and go.sum still appear in files changed

Comment (mlavacca) [reply go.mod:1]: yeah, you are right, they do. The changes are needed by the current implementation.

File: pkg/i2gw/providers/common/utils.go (hunk 1 above)
Comment (mlavacca) [reply pkg/i2gw/providers/common/utils.go]: oh I did not get your point initially. Yes, we can wrap it in a `trimsuffix`. Changed 👍

Comment (LiorLieberman) [reply pkg/i2gw/providers/common/utils.go]: so before this change, `routeName`used to be <ingressname>-all-hosts when`host==""`. and not it will be <ingressName> ? 
I dont mind just want to check I understand correctly and if we should not change other tests to align with this

Comment (LiorLieberman) [reply pkg/i2gw/providers/common/utils.go]: trimsuffix wont retain the same functionality, whats we want to have when host=="" ?

File: pkg/i2gw/providers/kong/converter.go (hunk 6 above)
Comment (LiorLieberman) [reply pkg/i2gw/providers/kong/converter.go:48]: any reason not to? We are copying the whole storage struct if we pass by value

File: pkg/i2gw/ingress2gateway.go [hunk 7]
```diff
@@ -127,3 +132,64 @@ func GetSupportedProviders() []string {
...
+			// 16 is the maximum number of addresses a Gateway can have
+			if len(g.Spec.Addresses) > 16 {
+				fieldPath := field.NewPath(fmt.Sprintf("%s/%s", nn.Namespace, nn.Name)).Child("spec").Child("addresses")
+				errs = append(errs, field.Invalid(fieldPath, g, "error while merging gateway listeners: a gateway cannot have more than 16 addresses"))
+			}
+		}
+	}
+
+	return newGateways, errs
+}
```
Comment (LiorLieberman) [inline pkg/i2gw/ingress2gateway.go:195]: Mind adding a comment that this behavior is likely to change following https://github.com/kubernetes-sigs/gateway-api/pull/1863/ ?

Comment (mlavacca) [reply pkg/i2gw/ingress2gateway.go]: added
